packages/core/src/index.ts

```diff
@@ -222,7 +222,23 @@ export const makeZodI18nMap: MakeZodI18nMap = (option) => (issue, ctx) => {
       );
       break;
     case ZodIssueCode.custom:
-      message = t(issue.params?.i18n ?? "errors.custom", {
+      const defaultKey = "errors.custom";
+      const i18nParam = issue.params?.i18n;
+      const { key, values } =
+        typeof i18nParam === "string"
+          ? { key: issue.params?.i18n, values: {} }
+          : typeof i18nParam === "object"
+          ? {
+              key: i18nParam?.key ?? defaultKey,
+              values:
+                typeof i18nParam?.values === "object"
+                  ? (i18nParam?.values as object)
+                  : {},
+            }
+          : { key: "errors.custom", values: {} };
+
+      message = t(key, {
+        ...values,
```

Comment:
I would like to avoid casting with `as object`.  
Please define the following functions outside of `makeZodI18nMap`:

```ts
const isRecord = (value: unknown): value is Record<string, unknown> => {
  if (typeof value !== "object" || value === null) return false;

  for (const key in value) {
    if (!Object.prototype.hasOwnProperty.call(value, key)) return false;
  }

  return true;
};

const getKeyAndValues = (
  param: unknown,
  defaultKey: string
): {
  values: Record<string, unknown>;
  key: string;
} => {
  if (typeof param === "string") return { key: param, values: {} };

  if (isRecord(param)) {
    const key =
      "key" in param && typeof param.key === "string" ? param.key : defaultKey;
    const values =
      "values" in param && isRecord(param.values) ? param.values : {};
    return { key, values };
  }

  return { key: defaultKey, values: {} };
};
```

Then, you can simplify the code and strictly check `values` by doing this:

```ts
case ZodIssueCode.custom:
      const { key, values } = getKeyAndValues(
        issue.params?.i18n,
        "errors.custom"
      );
      message = t(key, {
        ...values,
        ns,
        defaultValue: message,
        ...path,
      });
      break;
```